relocation.bs

```diff
@@ -1626,6 +1624,262 @@ auto [name, flag, nodes] = reloc tp;
+
+```c++
+struct MyClass
+{
+    gsl::non_null<std::unique_ptr<int>> _reloc_only;
+
+    auto get_all(this MyClass self)
+    {
+        return reloc self._reloc_only; // ill-formed, cannot relocate data members
+    }
+}
+```
+
+We introduce a special kind of member functions, called *decomposing functions*,
+to enable relocation of data-members. A decomposing function is a member
+function that takes an explicit `this`
+parameter by value and whose name is `reloc`:
+
+```c++
+auto get_all(this T reloc)
+```
+
+### Declaration ### {#decomposing-func-declare}
+
+The decomposing trait of a function may not appear in the declaration if the
+parameter name is omitted or different than `reloc`. The decomposing trait
+is an implementation detail and it does not need to appear in the declaration.
+
+A decomposing function does not bear any constraints on the number and types of
+parameters, except that the first one must be an explicit this parameter passed
+by value.
+
+### Definition ### {#decomposing-func-definition}
+
+A decomposing function decomposes its `this` parameter into subobjects before
+the function body is entered. It is subject to the following rules:
+
+- The explicit `this` object is inaccessible in the function body ;
+- Upon calling the function, before the function body is entered,
+    `*this` is decomposed into individual objects for each direct base and
+    non-static data member. This decomposition is a no-op: all subobjects of
+    `*this` merely get considered as distinct complete objects. In particular the
+    decomposition does not odr-use any of the copy, move or relocation constructor
+    of the subobjects.
+- Although the lifetime of `*this` ends once decomposed, its destructor is never
+    called ;
+- The direct base objects have no identifiers, and can only be accessed once
+    through the `reloc Base` expression (`Base` being a direct base of the
+    class-type). In particular, base data members are inaccessible in the
+    function body ;
+- The non-static data member objects are identified by their name in the class
+    declaration and can be accessed normally. If disambiguation is necessary then
+    the data members can be accessed with the syntax `T::datamember`, `T` being
+    the type of the `this` parameter, and `datamember` being the name of the
+    data member.
+
+Special considerations are taken for overlapping subobjects:
+
+- **EBO bases and** `[[no_unique_address]]` **members:** Obey to the above rules.
+- **Anonymous unions:** Obey to the above rules. `reloc` rules on function-local
+    anonymous union members apply.
+- **Virtual bases (and bases with virtual bases):** are unrelocatable. They cannot
+    be accessed by the `reloc Base` expression (ill-formed).
+
+As subobjects got turned into complete objects, they can be relocated like any
+function-local variable, or their destructor is otherwise called at scope exit.
```

Comment:
```suggestion
Since subobjects have become complete objects, they can be relocated like any
function-local variable; otherwise, their destructor is called on exit from function scope.
```